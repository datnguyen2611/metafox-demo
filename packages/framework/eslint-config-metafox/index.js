const restrictedGlobals = require('confusing-browser-globals');

module.exports = {
  root: true,
  // parser: '@babel/eslint-parser',
  parser: '@typescript-eslint/parser',
  extends: ['eslint:recommended'],
  plugins: [
    'import',
    'flowtype',
    'jsx-a11y',
    'react',
    'react-hooks',
    'jest-dom'
  ],

  env: {
    browser: true,
    commonjs: true,
    es6: true,
    jest: true,
    node: true
  },
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true
    }
  },

  settings: {
    react: {
      version: 'detect'
    }
  },
  overrides: [
    {
      files: ['**/*.ts?(x)'],
      // parser: '@typescript-eslint/parser',
      parserOptions: {
        ecmaVersion: 2020,
        sourceType: 'module',
        ecmaFeatures: {
          jsx: true
        },
        // typescript-eslint specific options
        warnOnUnsupportedTypeScriptVersion: true
      },
      plugins: ['@typescript-eslint'],
      // If adding a typescript-eslint version of an existing ESLint rule,
      // make sure to disable the ESLint rule here.
      rules: {
        // TypeScript's `noFallthroughCasesInSwitch` option is more robust (#6906)
        'default-case': 'off',
        // 'tsc' already handles this
        // (https://github.com/typescript-eslint/typescript-eslint/issues/291)
        'no-dupe-class-members': 'off',
        // 'tsc' already handles this
        // (https://github.com/typescript-eslint/typescript-eslint/issues/477)
        'no-undef': 'off',

        // Add TypeScript specific rules (and turn off ESLint equivalents)
        '@typescript-eslint/consistent-type-assertions': 'warn',
        'no-array-constructor': 'off',
        '@typescript-eslint/no-array-constructor': 'warn',
        'no-use-before-define': 'off',
        '@typescript-eslint/no-use-before-define': [
          'error',
          {
            functions: false,
            classes: false,
            variables: false,
            typedefs: false
          }
        ],
        'no-unused-expressions': 'off',
        '@typescript-eslint/no-unused-expressions': [
          'error',
          {
            allowShortCircuit: true,
            allowTernary: true,
            allowTaggedTemplates: true
          }
        ],
        'no-unused-vars': 'off',
        '@typescript-eslint/no-unused-vars': [
          'error',
          {
            args: 'none',
            ignoreRestSiblings: true
          }
        ],
        'no-useless-constructor': 'off',
        '@typescript-eslint/no-useless-constructor': 'warn'
      }
    }
  ],

  // NOTE: When adding rules here, you need to make sure they are compatible with
  // `typescript-eslint`, as some rules such as `no-array-constructor` aren't compatible.
  rules: {
    // http://eslint.org/docs/rules/
    'array-callback-return': 'warn',
    'default-case': [
      'warn',
      {
        commentPattern: '^no default$'
      }
    ],
    eqeqeq: ['warn', 'smart'],
    'new-parens': 'warn',
    'no-array-constructor': 'warn',
    'no-caller': 'warn',
    'no-cond-assign': ['warn', 'except-parens'],
    'no-const-assign': 'error',
    'no-control-regex': 'warn',
    'no-delete-var': 'warn',
    'no-dupe-args': 'warn',
    'no-dupe-class-members': 'error',
    'no-duplicate-imports': [
      2,
      {
        includeExports: false
      }
    ],
    'no-dupe-keys': 'warn',
    'no-duplicate-case': 'warn',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-eval': 'warn',
    'no-ex-assign': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-extra-label': 'warn',
    'no-fallthrough': 'warn',
    'no-func-assign': 'warn',
    'no-implied-eval': 'warn',
    'no-invalid-regexp': 'warn',
    'no-iterator': 'warn',
    'no-label-var': 'warn',
    'no-labels': [
      'warn',
      {
        allowLoop: true,
        allowSwitch: false
      }
    ],
    'no-lone-blocks': 'warn',
    'no-loop-func': 'warn',
    'no-mixed-operators': [
      'warn',
      {
        groups: [
          ['&', '|', '^', '~', '<<', '>>', '>>>'],
          ['==', '!=', '===', '!==', '>', '>=', '<', '<='],
          ['&&', '||'],
          ['in', 'instanceof']
        ],
        allowSamePrecedence: false
      }
    ],
    'no-multi-str': 'warn',
    'no-native-reassign': 'warn',
    'no-negated-in-lhs': 'warn',
    'no-new-func': 'warn',
    'no-new-object': 'warn',
    'no-new-symbol': 'warn',
    'no-new-wrappers': 'warn',
    'no-obj-calls': 'warn',
    'no-octal': 'warn',
    'no-octal-escape': 'warn',
    'no-redeclare': [
      'warn',
      {
        builtinGlobals: false
      }
    ],
    'no-regex-spaces': 'warn',
    'no-restricted-syntax': ['warn', 'WithStatement'],
    'no-script-url': 'warn',
    'no-self-assign': 'warn',
    'no-self-compare': 'warn',
    'no-sequences': 'warn',
    'no-shadow-restricted-names': 'warn',
    'no-sparse-arrays': 'warn',
    'no-template-curly-in-string': 'warn',
    'no-this-before-super': 'warn',
    'no-throw-literal': 'warn',
    'no-undef': 'error',
    'no-restricted-globals': ['error'].concat(restrictedGlobals),
    'no-unreachable': 'warn',
    'no-unused-expressions': [
      'error',
      {
        allowShortCircuit: true,
        allowTernary: true,
        allowTaggedTemplates: true
      }
    ],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        args: 'none',
        ignoreRestSiblings: true
      }
    ],
    'no-use-before-define': [
      'warn',
      {
        functions: false,
        classes: false,
        variables: false
      }
    ],
    'no-useless-computed-key': 'warn',
    'no-useless-concat': 'warn',
    'no-useless-constructor': 'warn',
    'no-useless-escape': 'warn',
    'no-useless-rename': [
      'warn',
      {
        ignoreDestructuring: false,
        ignoreImport: false,
        ignoreExport: false
      }
    ],
    'no-with': 'warn',
    'space-before-function-paren': [
      2,
      {
        anonymous: 'always',
        named: 'never',
        asyncArrow: 'always'
      }
    ],
    'template-curly-spacing': [2, 'never'],
    'nonblock-statement-body-position': 0,
    'no-whitespace-before-property': 'warn',
    'react-hooks/exhaustive-deps': 'warn',
    'require-yield': 'warn',
    'yield-star-spacing': [1, 'after'],
    'rest-spread-spacing': [1, 'never'],
    strict: ['warn', 'never'],
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'warn',
    'wrap-regex': 'off',
    'sort-imports': [
      0,
      {
        ignoreCase: false,
        ignoreDeclarationSort: false,
        ignoreMemberSort: true,
        memberSyntaxSortOrder: ['none', 'all', 'single', 'multiple'],
        allowSeparatedGroups: false
      }
    ],
    'valid-typeof': 'warn',
    'no-restricted-properties': [
      'error',
      {
        object: 'require',
        property: 'ensure',
        message:
          'More info: https://facebook.github.io/create-react-app/docs/code-splitting'
      },
      {
        object: 'System',
        property: 'import',
        message:
          'More info: https://facebook.github.io/create-react-app/docs/code-splitting'
      }
    ],
    'no-multi-spaces': [
      2,
      {
        ignoreEOLComments: true,
        exceptions: {
          Property: true
        }
      }
    ],
    'comma-spacing': [
      2,
      {
        before: false,
        after: true
      }
    ],
    'getter-return': 'warn',
    'import/first': 'error',
    'import/no-amd': 'error',
    'import/no-anonymous-default-export': 'warn',
    'import/no-webpack-loader-syntax': 'error',
    'react/forbid-foreign-prop-types': [
      'warn',
      {
        allowInPropTypes: true
      }
    ],
    'react/jsx-one-expression-per-line': 0,
    'react/jsx-no-literals': 'off',
    'react/jsx-props-no-multi-spaces': 'warn',
    'react/jsx-no-useless-fragment': 'warn',
    'react/jsx-no-script-url': 'error',
    'react/jsx-no-target-blank': 'warn',
    'react/jsx-no-comment-textnodes': 'warn',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-undef': 'error',
    'react/jsx-sort-props': [
      0,
      {
        callbacksLast: true,
        shorthandFirst: true,
        ignoreCase: true,
        noSortAlphabetically: false,
        reservedFirst: true
      }
    ],
    'react/jsx-pascal-case': [
      'warn',
      {
        allowAllCaps: true,
        ignore: []
      }
    ],
    'react/jsx-wrap-multilines': [
      2,
      {
        declaration: 'parens',
        assignment: 'parens',
        return: 'parens',
        arrow: 'parens',
        condition: 'ignore',
        logical: 'ignore',
        prop: 'ignore'
      }
    ],
    'react/jsx-uses-react': 'warn',
    'react/jsx-uses-vars': 'warn',
    'react/no-danger-with-children': 'warn',
    'react/no-direct-mutation-state': 'warn',
    'react/no-is-mounted': 'warn',
    'react/no-typos': 'error',
    'react/require-render-return': 'error',
    'react/no-string-refs': 'warn',
    'react/no-unsafe': 'warn',
    'react/no-this-in-sfc': 'warn',
    'react/no-unescaped-entities': 'warn',
    'react/no-unknown-property': [
      1,
      {
        ignore: []
      }
    ],
    'react/react-in-jsx-scope': 'error',
    'react/sort-prop-types': 'warn',
    'react/style-prop-object': 'warn',
    'react/forbid-dom-props': 'warn',
    'react/jsx-boolean-value': [1, 'never'],
    'react/jsx-equals-spacing': 'warn',
    'react/jsx-first-prop-new-line': 'warn',
    'react/jsx-child-element-spacing': 'warn',
    'react/jsx-closing-bracket-location': 0,
    'react/jsx-closing-tag-location': 0,
    'react/self-closing-comp': [
      0,
      {
        component: true,
        html: true
      }
    ],
    'react/jsx-max-props-per-line': [
      1,
      {
        maximum: 1,
        when: 'multiline'
      }
    ],
    'react/jsx-fragments': [1, 'syntax'],
    'jsx-a11y/alt-text': 'warn',
    'jsx-a11y/anchor-has-content': 'warn',
    'react/jsx-indent': 0,
    'react/jsx-indent-props': [1, 2],
    'react/jsx-key': 'warn',
    'react/jsx-max-depth': [
      1,
      {
        max: 15
      }
    ],
    'jsx-a11y/anchor-is-valid': [
      'warn',
      {
        aspects: ['noHref', 'invalidHref']
      }
    ],
    'jsx-a11y/aria-activedescendant-has-tabindex': 'warn',
    'jsx-a11y/aria-props': 'warn',
    'jsx-a11y/aria-proptypes': 'warn',
    'jsx-a11y/aria-role': [
      'warn',
      {
        ignoreNonDOM: true
      }
    ],

    // START eslint-recommended
    'no-console': 1,
    'no-constant-condition': 1,
    'no-debugger': 1,
    'no-empty': 1,
    'no-extra-boolean-cast': 1,
    'no-extra-semi': 1,
    'no-inner-declarations': 1,
    'no-irregular-whitespace': 1,
    'no-mixed-spaces-and-tabs': 1,
    // basic others
    indent: 0,
    'max-len': [
      2,
      120,
      4,
      {
        ignoreComments: true,
        ignoreUrls: true
      }
    ],
    'no-tabs': [
      1,
      {
        allowIndentationTabs: false
      }
    ],
    quotes: [2, 'single', { avoidEscape: true }],
    semi: [2, 'always'],
    'no-multiple-empty-lines': [
      2,
      {
        max: 1
      }
    ],
    'comma-dangle': [
      2,
      {
        arrays: 'never',
        objects: 'never',
        imports: 'never',
        exports: 'never',
        functions: 'never'
      }
    ],
    'padding-line-between-statements': [
      2,
      { blankLine: 'always', prev: 'if', next: '*' },
      { blankLine: 'always', prev: 'directive', next: '*' },
      { blankLine: 'always', prev: '*', next: ['return', 'if', 'function'] },
      { blankLine: 'always', prev: 'block-like', next: 'const' },
      {
        blankLine: 'always',
        prev: 'const',
        next: ['block-like', 'multiline-expression']
      },
      { blankLine: 'never', prev: 'import', next: 'import' },
      // { blankLine: 'never', prev: 'export', next: 'export' },
      {
        blankLine: 'always',
        prev: 'import',
        next: ['const', 'function', 'export', 'class']
      },
      { blankLine: 'always', prev: 'block', next: 'multiline-const' }
    ],
    'prefer-arrow-callback': [2],
    'object-shorthand': [2],
    'dot-location': [1, 'property'],
    'one-var': [2, 'never'],
    'no-var': [2],
    'prefer-template': [1],
    'prefer-const': ['error'],
    'prefer-destructuring': [0],
    'no-bitwise': [0],
    'id-length': [
      'error',
      {
        properties: 'never',
        exceptions: [
          'x',
          'y',
          'q',
          'i',
          'I',
          'e',
          'u',
          'n',
          'k',
          'a',
          'b',
          'c',
          's',
          't',
          '_',
          'v',
          'w',
          'h'
        ]
      }
    ],
    'func-names': [1, 'always'],
    'no-confusing-arrow': [
      1,
      {
        allowParens: true
      }
    ],
    yoda: 0,
    'object-curly-spacing': [2, 'always'],
    'array-bracket-spacing': 0,
    'no-class-assign': 'error',
    'space-unary-ops': [
      2,
      {
        words: false,
        nonwords: false
      }
    ],
    'switch-colon-spacing': [
      2,
      {
        after: true,
        before: false
      }
    ],
    'keyword-spacing': [
      'error',
      {
        after: true
      }
    ],
    'space-infix-ops': [
      2,
      {
        int32Hint: false
      }
    ],
    'space-before-blocks': [2, 'always'],
    'space-in-parens': [2, 'never'],
    'spaced-comment': [2, 'always', { markers: ['/'] }],
    'arrow-spacing': [
      2,
      {
        before: true,
        after: true
      }
    ],
    'arrow-parens': 0,
    'linebreak-style': [1, 'unix'],
    'operator-linebreak': 0,
    'function-paren-newline': 0,
    'react/prefer-stateless-function': 'off',
    'import/newline-after-import': [0],
    'react/jsx-filename-extension': [0],
    'react/forbid-prop-types': [0],
    'import/no-named-as-default': [0],
    'import/no-named-as-default-member': [0],
    'import/no-extraneous-dependencies': [0],
    'import/no-unresolved': [0],
    'import/extensions': [0],
    // END eslint-recommended
    'jsx-a11y/aria-unsupported-elements': 'warn',
    'jsx-a11y/heading-has-content': 'warn',
    'jsx-a11y/iframe-has-title': 'warn',
    'jsx-a11y/img-redundant-alt': 'warn',
    'jsx-a11y/no-access-key': 'warn',
    'jsx-a11y/no-distracting-elements': 'warn',
    'jsx-a11y/no-redundant-roles': 'warn',
    'jsx-a11y/role-has-required-aria-props': 'warn',
    'jsx-a11y/role-supports-aria-props': 'warn',
    'jsx-a11y/scope': 'warn',
    'react-hooks/rules-of-hooks': 'error',
    'flowtype/define-flow-type': 'warn',
    'flowtype/require-valid-file-annotation': 'warn',
    'flowtype/use-flow-type': 'warn',

    // eslint-plugin-jest-dom
    // https://github.com/testing-library/eslint-plugin-jest-dom
    'jest-dom/prefer-checked': 'error',
    'jest-dom/prefer-enabled-disabled': 'error',
    'jest-dom/prefer-required': 'error',
    'jest-dom/prefer-to-have-attribute': 'error'
  }
};
