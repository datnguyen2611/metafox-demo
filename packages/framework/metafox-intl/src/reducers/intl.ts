/**
 * @type: reducer
 * name: intl
 */
import { createAction, createReducer } from '@reduxjs/toolkit';
import { AppState } from '@metafox/intl/types';

const setMessages = createAction('intl/setMessages');

export default createReducer<AppState>(
  {
    messages: {}
  },
  builder => {
    builder.addCase(setMessages, (state, action) => {
      state.messages = action.payload;
    });
  }
);
