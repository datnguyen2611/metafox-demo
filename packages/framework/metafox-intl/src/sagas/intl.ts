/**
 * @type: saga
 * name: metafox.intl
 */

import {
  APP_BOOTSTRAP,
  getGlobalContext,
  GlobalState,
  LOCALE_GROUP,
  IS_ADMINCP,
  USE_BOOTSTRAP_CACHE as CACHED
} from '@metafox/framework';
import { get } from 'lodash';
import { put, select, takeLeading } from 'redux-saga/effects';

const CACHED_PHRASE_KEY = IS_ADMINCP ? 'phrases.admin' : 'phrases';

export function* bootstrapIntl() {
  // load locale data/

  const { apiClient, localStore, preferenceBackend } =
    yield* getGlobalContext();

  try {
    let data: Record<string, string> = CACHED
      ? localStore.getJSON(CACHED_PHRASE_KEY)
      : undefined;

    const revision = CACHED ? get(data, 'revision', 'now') : 'now';

    const defaultMessages = yield select(
      (state: GlobalState) => state.intl.messages
    );

    const response = yield apiClient.get(
      `core/translation/${LOCALE_GROUP}/auto/${revision}`
    );

    const keepCached = get(response, 'data.data.keepCached', false);
    const $locale = get(response, 'data.data.$locale');

    if ($locale) {
      preferenceBackend.set('userLanguage', $locale);
      // handle locale changed.
    }

    if (!keepCached) {
      data = response.data.data;
      localStore.set(CACHED_PHRASE_KEY, JSON.stringify(data));
    }

    yield put({
      type: 'intl/setMessages',
      payload: { ...defaultMessages, ...data }
    });
    yield put({ type: '@bootstrap/intl/DONE' });
  } catch (err) {
    yield put({ type: '@bootstrap/intl/DONE', error: true });
  }
}

const sagas = [takeLeading(APP_BOOTSTRAP, bootstrapIntl)];

export default sagas;
