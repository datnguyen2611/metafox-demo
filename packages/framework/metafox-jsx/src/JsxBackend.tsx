/**
 * @type: service
 * name: jsxBackend
 */
import { RenderBaseItem, RenderItemProps } from '@metafox/framework';
import { isString } from 'lodash';
import React from 'react';
import { isValidElementType } from 'react-is';
import { JsxBackendConfig } from './types';

/**
 * JsxBackend provider virtual component map by name => element type
 *
 */
export default class JsxBackend {
  /**
   * Configuration key.
   */
  public static readonly configKey: string = 'views';

  /**
   * store all views
   */
  private data: JsxBackendConfig;

  /**
   * default constructor
   * @constructor
   */
  constructor(data: JsxBackendConfig) {
    this.data = data;
    this.use = this.use.bind(this);
  }

  /**
   * @internal
   * this method is private and called by Manager
   */
  public bootstrap() {}

  /**
   * Extending current configure
   *
   * @param {JsxBackendConfig} opts - a dictionary of Jsx Component
   */
  public use(opts: JsxBackendConfig) {
    Object.keys(opts).forEach(key => {
      this.data[key] = opts[key];
    });
  }

  public get<T = any>(
    tagName: string | React.ElementType
  ): React.ElementType<T> {
    return isString(tagName) && this.data[tagName]
      ? this.data[tagName]
      : undefined;
  }

  public find(fn: (key: string) => boolean): string[] {
    return Object.keys(this.data).filter(fn);
  }

  /**
   * Get all jsx component map
   * @returns
   */
  public getAll(): JsxBackendConfig {
    return this.data;
  }

  /**
   *
   * @param tagName - name of component or tags
   * @returns boolean
   */
  public has(tagName: any): boolean {
    return 'string' === typeof tagName && this.data[tagName] !== undefined;
  }

  /**
   *
   * @param {RenderItemProps} item  - array or single render item
   * @param {Function} filter (optional) - filter
   * @returns -  Jsx Element
   */
  public render(
    item: RenderItemProps,
    filter?: (c: RenderBaseItem, index: number) => boolean
  ) {
    if (Array.isArray(item)) {
      if (filter) {
        return item.filter(filter).map(c => this.render(c));
      } else {
        return item.map(c => this.render(c)).filter(Boolean);
      }
    } else if (item?.component) {
      let { component: tag } = item;
      const { props } = item;

      if ('string' === typeof tag && this.data[tag]) {
        tag = this.data[tag];
      } else if (!isValidElementType(tag)) {
        return null;
      } else if ('string' === typeof tag) {
        if (!/div|span|p|h1|h2|h3|h4|h5|h6|b|i|img/i.test(tag)) return null;
      }

      return React.createElement(tag, props);
    }

    return null;
  }
}
