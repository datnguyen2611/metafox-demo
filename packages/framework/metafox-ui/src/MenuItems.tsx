import { MenuShape, useGlobal } from '@metafox/framework';
import { MenuItemViewProps } from '@metafox/ui';
import { assign } from 'lodash';

export type Props = Pick<
  MenuItemViewProps,
  'pathname' | 'classes' | 'handleAction'
> & {
  items: MenuShape['items'];
  prefixName?: string;
  fallbackName?: string;
  identity?: string;
  closeMenu(): void;
};

export default function MenuItems({
  prefixName = 'menuItem.as.',
  fallbackName = 'normal',
  identity,
  items,
  closeMenu,
  ...rest
}: Props): JSX.Element {
  const { jsxBackend } = useGlobal();

  return items.map((item, index) =>
    jsxBackend.render({
      component: `${prefixName}${item.as || fallbackName}`,
      props: assign({ key: index, item, identity, closeMenu }, rest)
    })
  ) as any;
}
