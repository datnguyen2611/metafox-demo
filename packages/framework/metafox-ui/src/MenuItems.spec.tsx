import { GlobalProvider, Manager } from '@metafox/framework';
import { IntlProvider } from '@metafox/intl';
import JsxBackend from '@metafox/jsx';
import { ThemeProvider } from '@metafox/layout';
import { render } from '@testing-library/react';
import * as React from 'react';
import MenuItems from './MenuItems';

const MockInfoLoader = () => <div data-testid="infoLoader" />;

const MockMoreMenu = () => <div data-testid="moreMenu">more menu</div>;
const MockDivider = () => <div data-testid="divider">divider</div>;

describe('MenuItems', () => {
  const items = [
    {
      as: 'popover',
      testid: 'more menu',
      icon: 'ico-dottedmoreo',
      label: 'More',
      name: 'more_menu',
      ordering: 1
    },
    {
      as: 'divider',
      testid: 'divider',
      icon: '',
      label: null,
      name: 'divider',
      ordering: 2
    }
  ];

  const jsxBackend = new JsxBackend({
    'menuItem.as.popover': MockMoreMenu,
    'menuItem.as.divider': MockDivider
  });

  const manager = Manager.factory({});
  const usePreference = jest.fn().mockImplementation(() => ({}));
  const getConfig = jest.fn().mockImplementation(() => {});
  const mockOn = jest.fn();
  const mockOff = jest.fn();
  const mockRenderFn = jest.fn(({ component, props }) => {
    if (component == 'menuItem.as.popover') {
      return jsxBackend.render({ component: 'menuItem.as.popover', props });
    }

    if (component == 'menuItem.as.divider') {
      return jsxBackend.render({ component: 'menuItem.as.divider', props });
    }
  });

  manager.use({
    usePreference,
    eventCenter: { on: mockOn, off: mockOff },
    constants: { isLayoutPreviewWindow: false },
    getConfig,
    jsxBackend: { render: mockRenderFn }
  });

  it('render', async () => {
    const { container, getByTestId } = render(
      <GlobalProvider value={manager}>
        <IntlProvider>
          <ThemeProvider infoLoader={MockInfoLoader}>
            <MenuItems items={items} />
          </ThemeProvider>
        </IntlProvider>
      </GlobalProvider>
    );
    expect(getByTestId('moreMenu')).toBeInTheDocument();
    expect(getByTestId('divider')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
