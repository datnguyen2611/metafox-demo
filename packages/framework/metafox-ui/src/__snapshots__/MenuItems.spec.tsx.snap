// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItems render 1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <div
    data-testid="moreMenu"
  >
    more menu
  </div>
  <div
    data-testid="divider"
  >
    divider
  </div>
</div>
`;
